### ISSUE TITLE
Proposal: Support authentication based on cookie containing encrypted session info

### ISSUE THREAD

**Issue by haslersn** [ory/oathkeeper#509]
## The problem

1. I'd like to authenticate users via OpenID Connect (OIDC). In principle this can be done by using Oathkeeper's `cookie_session` authenticator and pointing the `check_session_url` to an instance of [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/).
2. Aforementioned `check_session_url` endpoint returns an access token (if oauth2-proxy is configured to do so) and I'd like to pass that access token to the upstream application (with Oathkeeper in reverse proxy mode) or to the downstream reverse proxy (with Oathkeeper in decision mode). I think this is not possible with Oathkeeper, yet, but doesn't seem too hard to implement. This is **not** the concern of this feature proposal. Keep reading for my concern.
3. Depending on the requested host (`Host` header with Oathkeeper in reverse proxy mode, or host contained in the `X-Original-URL` header with Oathkeeper in decision mode), I'd like to use host-specific OIDC client credentials and scope, so that the access token has only the scopes needed by that specific host. Two nice side effects would be that (a) I don't need to use a wildcard cookie domain and (b) the user has to authorize each host individually.

Unfortunately oauth2-proxy does not support host-dependent configuration of OIDC client credentials and scope. I could use one oauth2-proxy instance per host, but then I'd have configuration in two places that needs to be kept in sync. That is, for each host, I'd need to configure an Oathkeeper rule that uses the correct oauth2-proxy instance and I'd need to configure that instance. This seems very cumbersome.

## Preliminaries

In order to understand my proposal, you should understand the following about how oauth2-proxy works:

1. oauth2-proxy has a login endpoint that initiates an authorization code grant flow where oauth2-proxy is the OIDC client. The OIDC provider is external and can be configured. When the user is authenticated, oauth2-proxy returns a cookie containing the encrypted session data (i.e. in particular the ID token and access token). Only the oauth2-proxy knows the decryption key. **No** session data is stored server-side.
2. When aforementioned `check_session_url` endpoint of oauth2-proxy is accessed, then 200 is returned only if the cookie is set and contains an encrypted ID token that is valid (i.e. signed by the OIDC provider).

## High-level proposal

To solve the problem, we could implement a new (built-in) Oathkeeper authenticator that replaces oauth2-proxy in above example.  Now I don't think most people would like Oathkeeper to become an OIDC Client, because that shouldn't be the concern of an identity and access proxy (IAP). However, we could implement only the second part (i.e. the check) in an Oathkeeper authenticator and implement the first part (i.e. the OIDC client) in a separate service (which I call "OpenID Connect commissioned client" (oidc3) from now on). Oathkeeper and oidc3 then simply use a shared secret, so that Oathkeeper can verify the content of cookies set by oidc3.

Note that the part that needs to be implemented in Oathkeeper is extremely lightweight.

Do you like this idea? Or do you have alternative proposals on how I can solve my problem?

### oidc3

oidc3 would then be a new software, which I'd be willing to develop and which shall have the following properties:

* Less bloated than oauth2-proxy, because oidc3 does not need to be a reverse proxy. It only needs to implement the OIDC client part.
* Supports host-dependent OIDC client credentials and scope.
* Configured in YAML instead of an undocumented cfg file format.

## Other advantages

In deployments where host-dependent configuration of oidc3 is not needed, my proposal still has the advantage that authentication happens directly in Oathkeeper and no extra HTTP request has to be made (if the OIDC provider's public key is cached). This is possible because only the shared secret and the OIDC provider's public key is needed in order to authenticate a session cookie.

In order to also have this advantage with existing software (and not only with oidc3), we should maybe use a session format compatible with oauth2-proxy. I'm going to look into which encryption algorithm and which format they use, in order to assess whether it's sane to stay compatible.

**Comment by aeneasr** [comment]
Thank you for the write up! Just to double-check, given that you have oidc3 as a server, wouldn't it be enough to use the `check_session` endpoint? Or is it necessary for ORY Oathkeeper to understand the oidc3 cookie?

**Comment by haslersn** [comment]
@aeneasr For that to be enough

- Oathkeeper needs to pass the `X-Original-URL` header to the `check_session` endpoint. (Currently not supported.)
- Oathkeeper needs to read the Authorization header from the `check_session` endpoint's response. (Currently not supported.)
- When running Oathkeeper in decision mode, it needs to return the Authrorization header downstream instead of upstream.

I'm not sure if the `header` mutator can currently pass headers downstream when Oathkeeper runs in decision mode. Can it?

**Comment by aeneasr** [comment]
Assuming that you would create a new endpoint which implements this, could you not also return that in the JSON response? Or is that specifically a limitation of the Bitly OAuth2 Proxy which you'd like to respect?

**Comment by haslersn** [comment]
@aeneasr You understood it correctly. If I don't want to stay compatible with oauth2-proxy, then I could return it as JSON.

Does anybody know what's the behaviour of the Oathkeeper `header` mutator when Oathkeeper runs in decision mode?

**Comment by aeneasr** [comment]
The header mutator simply changes the headers of the JSON response in that case!

Regarding the bitly server, I'm not sure if we really want to support that out of the box, given that last time I checked it looked pretty inactive and also quite buggy.

**Comment by jakhog** [comment]
Hi @haslersn and @aeneasr. We are currently working on a similar scenario where we would like to authenticate with OpenID/OAuth, and store the session in a cookie. As @haslersn described, the verification of the cookie is easy with Oathkeeper. But we also need to create and store the session cookie.

The way we have started doing this right now is that we are building a separate service (like the proposed oidc3), that is setup as a redirect error handler in Oathkeeper, which acts as an OIDC client and stores the received token in a cookie. We have not spent a lot of time thinking about the contents of the cookie yet, but we were thinking of using the _access token_ (from Hydra) as the cookie, and then using the token introspection or userinfo endpoints during authentication in Oathkeeper.

It is however very little code needed to do this flow, and it would be nice to have a seamless integration between Oathkeeper and Hydra. So, my question is, is it actually a bad idea to enable Oathkeeper to be an OpenID/OAuth client? If not, my idea would be to do this as an error handler, and then be able to configure the provider and what token to put in the cookie. Looking at the future NextGen plans for Oathkeeper, it seems like there are thoughts on making more configurable error handlers which could allow per-host ID providers with this approach.

Is this something you could want in Oathkeeper @aeneasr, or should I join forces with @haslersn on the oidc3 proposal?

**Comment by aeneasr** [comment]
It depends a bit, the problem with your approach is that you would probably also need a UI where the user could select a login provider. That would imply that we need some type of UI in this project unless we only allow a single oidc provider to handle the login. I imagine what you want would probably have feature parity with https://github.com/pomerium/pomerium ?

**Comment by haslersn** [comment]
I think in the majority of cases this is about integrating Oathkeeper with Hydra, so there's only a single OIDC provider required.

For the use case if somebody wants to allow for choosing between multiple (usually external) OIDC providers, I have the following idea: You can deploy a Hydra instance and use it as a proxy. That is:

1. You configure Oathkeeper to use that Hydra instance as OIDC provider
2. You program a new identity provider for Hydra that lets the user choose between several other OIDC providers, acts as an OIDC client towards the chosen OIDC provider and proxies the result back to Hydra.

**Comment by jakhog** [comment]
Yes, for us the case is also integrating Oathkeeper with Hydra (or any other OIDC/OAuth2 provider). In our case we also want to allow multiple external IDPs, but for that we are using Kratos - with our custom identity model. So Hydra is issuing tokens with identities from Kratos.

This works great, and we also want to use Hydra as token issuer for other non-browser applications, which is very easy.

My goal for what I'm thinking about in Oathkeeper is to somewhat emulate an OIDC Client with the browser. So what we need to to is just initiate a token request, and store the result (somehow) in a cookie. On the surface it seems like a good idea, but I might have missed a lot of things related to encrypting values in cookies, and refreshing tokens, so maybe it is more involved than I imagined.

**Comment by aeneasr** [comment]
Right! I think we can take this into consideration for Ory Oathkeeper Next! Another idea would be using query params to select an authenticator. That way you could easily show a UI (could even be an SPA) to select the right provider.
